src/NServiceBus.Core/Sagas/ActiveSagaInstance.cs

```diff
@@ -222,9 +210,22 @@ static object GetDefault(Type type)
 
         internal class CorrelationPropertyInfo
         {
-            public bool HasExistingValue;
-            public PropertyInfo PropertyInfo;
-            public object Value;
+            public CorrelationPropertyInfo(PropertyInfo propertyInfo, IContainSagaData sagaData)
+            {
+                this.sagaData = sagaData;
+
+                PropertyInfo = propertyInfo;
+                InitialValue = propertyInfo.GetValue(sagaData);
+                var defaultValue = GetDefault(propertyInfo.PropertyType);
+                HasInitialValue = InitialValue != null && !InitialValue.Equals(defaultValue);
+            }
+
+            IContainSagaData sagaData;
+
+            public bool HasInitialValue { get; }
+            public PropertyInfo PropertyInfo { get; }
+            public object InitialValue { get; }
+            public object CurrentValue => PropertyInfo.GetValue(sagaData);
```

Comment:
would it make sense to add a comment here that calling this property multiple times is expensive due to internal reflection usage? I guess caching it on first access might be dangerous? Mentioning it because it seems that we are now using reflection three times instead of once. Namely

```
correlationProperty.InitialValue.ToString() == correlationProperty.CurrentValue.ToString()
```

and

```
sagaCorrelationProperty = new SagaCorrelationProperty(correlationProperty.PropertyInfo.Name, correlationProperty.CurrentValue);
```

and in the exception block (not worried about that since this is expensive path anyway)